colossalai/inference/executor/rpc_worker.py

```diff
@@ -0,0 +1,294 @@
+                    device=get_accelerator().get_current_device(),
+                )
+            )
+        logger.info("physical cache init over")
+
+    def exposed_execute_model_forward(self, input_token_ids_param: List[int], input_meta_data_param: dict):
+        # prepare the data for model forward
+        input_meta_data = InputMetaData.from_rpc_param(input_meta_data_param)
+        input_meta_data.fd_inter_tensor = self.fd_inter_tensor
+        if input_meta_data.is_prompts:
+            n_tokens = input_meta_data.sequence_lengths.sum().item()
+        else:
+            n_tokens = input_meta_data.batch_size
+        input_token_ids = torch.tensor(input_token_ids_param, dtype=torch.int, device=self.device)
+
+        # execute the model
+        logits = self.model(
+            input_token_ids,
+            self.output_tensor[:n_tokens],
+            input_meta_data,
+            self.k_cache,
+            self.v_cache,
+        )
+
+        # sampler
+        if self.inference_config.pad_input:
+            logits = logits[:, -1, :]
+        next_tokens = search_tokens(
+            self.inference_config.to_generation_config(self.model_config), logits, input_meta_data.is_prompts
+        )
+
+        # return the tokens generated to scheduler
+        return next_tokens.tolist()
+
+    def _init_output_tensor(self):
+        alloc_shape = (
+            self.inference_config.max_batch_size
+            * (self.inference_config.max_input_len + self.inference_config.max_output_len),
+            self.model_config.hidden_size // self.inference_config.tp_size,
+        )
+        self.output_tensor = torch.zeros(alloc_shape, dtype=self.dtype, device=self.device)
+
+    def _init_fd_tensor(self):
+        fd_inter_tensor = FDIntermTensors()
+
+        if fd_inter_tensor._tensors_initialized:
+            fd_inter_tensor._reset()
+
+        # For Spec-Dec, process the speculated tokens plus the token in the last step for each seq
+        max_n_tokens = self.inference_config.max_batch_size
+        max_n_tokens *= self.inference_config.max_n_spec_tokens + 1
+
+        inference_config = self.inference_config
+        kv_max_split_num = (
+            inference_config.max_input_len + inference_config.max_output_len + inference_config.block_size - 1
+        ) // inference_config.block_size
+        head_dim = self.model_config.hidden_size // self.model_config.num_attention_heads
+
+        fd_inter_tensor.initialize(
+            max_batch_size=max_n_tokens,
+            num_attn_heads=self.model_config.num_attention_heads // self.inference_config.tp_size,
+            kv_max_split_num=kv_max_split_num,
+            head_dim=head_dim,
+            dtype=self.dtype,
+            device=get_accelerator().get_current_device(),
+        )
+
+        self.fd_inter_tensor = fd_inter_tensor
+
+    def _init_model(self, model_or_path: Union[nn.Module, str], model_policy: Policy = None):
+        """
+        Shard model or/and Load weight
```

Comment:
We shall describe the conditions to make different choices in docstring

Reply:
Thx, done.